zebra-state/src/service/chain_tip.rs

```diff
@@ -509,6 +524,94 @@ impl ChainTipChange {
         Ok(action)
     }
 
+    /// Returns a channel that passes any new blocks that are added the the best chain tip that it
+    /// hasn't already sent.
+    pub fn spawn_wait_for_blocks(mut self) -> tokio::sync::mpsc::UnboundedReceiver<Arc<Block>> {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+
+        tokio::spawn(async move {
+            // # Correctness
+            //
+            // This task should see any `ChainTipBlocks` sent by the block write task through
+            // `ChainSender` because  `wait_for_blocks()` should be able to process blocks faster
+            // than they can be verified.
+            while let Ok(blocks) = tokio::select! {
+                blocks = self.wait_for_blocks() => blocks,
+                _ = tx.closed() => { return; }
+            } {
+                for block in blocks {
+                    if tx.send(block).is_err() {
+                        return;
+                    }
+                }
+            }
+        });
+
+        rx
+    }
+
+    /// Returns the new best chain tip when its previous block hash matches the last_change_hash,
+    /// or any blocks in the new best chain that it hasn't returned already otherwise.
+    ///
+    /// # Panics
+    ///
+    /// This method is meant to be used with `spawn_wait_for_blocks` and could panic when called
+    /// after the `ChangeTipChange` instance misses a tip block change.
+    async fn wait_for_blocks(&mut self) -> Result<Vec<Arc<Block>>, watch::error::RecvError> {
+        // Get the previous last_change_hash before `wait_for_tip_change()` updates it.
+        let last_change_hash = self.last_change_hash;
+        let block = self.tip_block_change().await?;
+
+        assert!(
+            Some(block.hash) != self.last_change_hash,
+            "ChainTipSender and ChainTipChange ignore unchanged tips"
+        );
+
+        self.last_change_hash = Some(block.hash);
+
+        // Return initial tip block
+        let Some(last_change_hash) = last_change_hash else {
+            return Ok(vec![block.block]);
+        };
+
+        // Return block if previous block hash matches last tip block hash
+        // TODO: Add a new variant for NetworkActivation upgrades?
+        if block.previous_block_hash() == last_change_hash {
+            return Ok(vec![block.block]);
+        }
+
+        // There's a new best chain
+
+        let non_finalized_state = self.non_finalized_state_receiver.cloned_watch_data();
```

Comment:
This may need a correctness comment about it relying on the block write task sending the updated non-finalized state before updating the chain tip sender.